x/config/dns.go

```diff
@@ -23,14 +24,69 @@ import (
 
 	"github.com/Jigsaw-Code/outline-sdk/dns"
 	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"golang.org/x/net/dns/dnsmessage"
 )
 
-func wrapStreamDialerWithDOH(innerDialer transport.StreamDialer, configURL *url.URL) (transport.StreamDialer, error) {
+func wrapStreamDialerWithDO53(innerSD func() (transport.StreamDialer, error), innerPD func() (transport.PacketDialer, error), configURL *url.URL) (transport.StreamDialer, error) {
+	sd, err := innerSD()
+	if err != nil {
+		return nil, err
+	}
+	pd, err := innerPD()
+	if err != nil {
+		return nil, err
+	}
 	query := configURL.Opaque
 	values, err := url.ParseQuery(query)
 	if err != nil {
 		return nil, err
 	}
+	var address string
+	for key, values := range values {
+		switch strings.ToLower(key) {
+		case "address":
+			if len(values) != 1 {
+				return nil, fmt.Errorf("address option must has one value, found %v", len(values))
+			}
+			address = values[0]
+		default:
+			return nil, fmt.Errorf("unsupported option %v", key)
+
+		}
+	}
+	if address == "" {
+		return nil, errors.New("must set an address")
+	}
+	_, _, err = net.SplitHostPort(address)
+	if err != nil {
+		address = net.JoinHostPort(address, "53")
+	}
+	udpResolver := dns.NewUDPResolver(pd, address)
+	tcpResolver := dns.NewTCPResolver(sd, address)
+	resolver := dns.FuncResolver(func(ctx context.Context, q dnsmessage.Question) (*dnsmessage.Message, error) {
```

Comment:
@fortuna this basically uses the inner packet dialer to do dns resolution over udp first, and if it fails, it uses inner stream dialer to do resolution over tcp right?

Reply:
No, it only falls back to TCP if the response is truncated.
It will fail if UDP fails.